### DISCUSSION TITLE
Is there a way to hook a function to transform data prior to validation process?

### DISCUSSION THREAD

**Discussion by angelhodar** [react-hook-form/react-hook-form#8090] [Q&A]
Hey! So my question is very simple. Doest react-hook-form allows to pass a function to transform the data that a a controlled input should return prior to submitting? The use case would be to return null instead of empty string in a select input, because I want the default value to be null instead of empty string. That way i can match the server validation exactly the same instead of changing yup fields to empty string when in server its null. I tried to set onChange in the select field with a setValue, but it doesnt allow to put null as its not supported by native web inputs.

```
onChange={(e) =>
          setValue(name, e.target.value === "" ? null : e.target.value)
        }
```

Thanks in advance!

**Comment by Moshyfawn** [comment]
It's advised not to mix the controller's value handlers _(`onChange`, `onBlur`, etc.)_ and the `setValue` form method.

Usually, for controlled inputs the value transformation process looks like
```jsx
<Controller
  control={control}
  name="fieldName"
  render={({ field: { onChange, ...fieldProps } }) => (
    <input {...fieldProps} onChange={event => onChange(event.target.value || null)} />
  )}
/>
```
So, you transform your new value and pass it into your field `onChange` handler.

You can create an abstraction like:
```js
const transform = (value: string) => value || null
...
onChange={event => field.onChange(transform(event.target.value))}
```

**Comment by Moshyfawn** [comment]
As you're using Yup, you can also look into [per-processing _(transforming)_ the input's value before validation](https://github.com/jquense/yup#parsing-transforms)

**Comment by angelhodar** [comment]
Thank you so much @Moshyfawn! My mistake was to not call the onChange that I received in the field object. The thing is that this solution actually sets the value to null in the input prop and the console throws an error because input elements cant accept null values. So my idea instead of onChange would be something like onPreSubmit or something like that to not apply the transformed value into the input component itself.
